File: FrockRaytracer/Raytracer.cs
using System;
using System.Collections.Generic;
using System.Threading;
using FrockRaytracer.Objects;
using FrockRaytracer.Objects.Primitives;
using FrockRaytracer.Structs;
using FrockRaytracer.Utils;
using OpenTK;
using OpenTK.Graphics.ES11;

namespace FrockRaytracer
{
    public class Raytracer
    {
        public World World { get; private set; }
        public DebugData DebugData { get; private set; } = new DebugData();

        public void Reset(World world)
        {
            World = world;
            DebugData.Clear();
            World.Cache();
        }
        
        /// <summary>
        /// Trace the ray and go around the world in 80ns. Or more :S
        /// </summary>
        /// <param name="ray"></param>
        /// <param name="debug"></param>
        /// <returns></returns>
        public Vector3 TraceRay(Ray ray, bool debug = false)
        {
            Vector3 ret = Vector3.Zero;
            if (ray.Depth > Settings.MaxDepth) return ret;
            
            // Check for intersections
            RayHit hit = intersect(ray);
            if (hit.Obj == null) return World.Environent == null ? Vector3.Zero : World.Environent.GetColor(ray);
            if (debug) DebugData.PrimaryRays.Add(new Tuple<Ray, RayHit>(ray, hit));
            
            // Calculate color and specular highlights. Pure mirrors dont have diffuse
            Vector3 specular = Vector3.Zero;
            Vector3 color = Vector3.Zero;
            if (!hit.Obj.Material.IsMirror) {
                color += illuminate(ray, hit, ref specular, debug);
                color += World.Environent.AmbientLight;
            }
            
            
            // Different materials are handled differently. Would be cool to move that into material
            if (hit.Obj.Material.IsMirror) {
                ret = TraceRay(RayTrans.Reflect(ray, hit), debug);
            } else if (hit.Obj.Material.IsDielectic) {
                var n1 = ray.isOutside() ? Constants.LIGHT_IOR : hit.Obj.Material.RefractionIndex; // TODO: shorten this shizzle into a func
                var n2 = ray.isOutside() ? hit.Obj.Material.RefractionIndex : Constants.LIGHT_IOR;
                float reflect_multiplier = QuickMaths.Fresnel(n1, n2, hit.Normal, ray.Direction);
                reflect_multiplier = hit.Obj.Material.Reflectivity + (1f - hit.Obj.Material.Reflectivity) * reflect_multiplier;
                float transmission_multiplier = 1 - reflect_multiplier;

                // Reflect if its worth it
                if (reflect_multiplier > Constants.EPSILON)
                {
                    Ray reflRay = RayTrans.Reflect(ray, hit);
                    uint takeSamples = 0;
                    if (hit.Obj.Material.Roughness > 0.001f) takeSamples = Settings.MaxReflectionSamples;
                    else takeSamples = 1;
                    for (int i = 0; i < takeSamples; i++)
                    {
                        Ray localRay = reflRay;
                        localRay.Direction = RRandom.RandomChange(reflRay.Direction, hit.Obj.Material.Roughness);
                        ret += reflect_multiplier * TraceRay(localRay, debug);
                    }
                    ret /= takeSamples;
                }
                if (transmission_multiplier > Constants.EPSILON) {
                    if(hit.Obj.Material.IsRefractive) {
                        var tmp_ray = RayTrans.Refract(ray, hit);
                        if(!float.IsNaN(tmp_ray.Direction.X)) // Happens rarely
                            ret += transmission_multiplier * traceRayInside(tmp_ray, hit.Obj, debug);
                    } else {
                        ret += transmission_multiplier * color;
                    }
                }
            } else {
                // Standard diffuse
                ret = color;
            }

            return ret + specular;
        }

        /// <summary>
        /// Traces the ray if its inside an object which means its probably in refraction loop
        /// </summary>
        /// <param name="ray"></param>
        /// <param name="obj"></param>
        /// <param name="debug"></param>
        /// <returns></returns>
        protected Vector3 traceRayInside(Ray ray, Primitive obj, bool debug)
        {
            if (ray.isOutside()) return TraceRay(ray, debug); // Should not happen anyway
            Vector3 ret = Vector3.Zero;
            float multiplier = 1;
            float absorb_distance = 0;

            // Do the refraction loop. Aka refract, reflect, repeat
            RayHit hit = RayHit.Default();
            while (ray.Depth < Settings.MaxDepth - 1 && multiplier > Constants.EPSILON) {
                if (!obj.Intersect(ray, ref hit)) return ret; // Should not happen either
                if(debug) DebugData.RefractRays.Add(new Tuple<Ray, RayHit>(ray, hit));

                // Beer absorption
                absorb_distance += hit.T;
                var absorb = QuickMaths.Exp(-obj.Material.Absorb * absorb_distance);

                // Fresnel
                var reflect_multiplier = QuickMaths.Fresnel(obj.Material.RefractionIndex, Constants.LIGHT_IOR, ray.Direction, hit.Normal);
                var refract_multiplier = 1f - reflect_multiplier;

                // refract if its worth
                if (refract_multiplier > Constants.EPSILON)
                    ret += TraceRay(RayTrans.Refract(ray, hit), debug) * refract_multiplier * multiplier * absorb;

                ray = RayTrans.Reflect(ray, hit);
                multiplier *= reflect_multiplier;
            }

            return ret;
        }

        /// <summary>
        /// Get diffuse color and specular for the hit point with a specific light
        /// </summary>
        /// <param name="light"></param>
        /// <param name="ray"></param>
        /// <param name="hit"></param>
        /// <param name="specular"></param>
        /// <param name="debug"></param>
        /// <returns></returns>
        protected Vector3 illuminateBy(Vector3 ppos, Light light, Ray ray, RayHit hit, ref Vector3 specular, bool debug)
        {
            // Calclulate if we are within light range
            var light_vec = ppos - hit.Position;
            float dist_sq = Vector3.Dot(light_vec, light_vec);
            float intens_sq = light.LightCache.IntensitySq;
            if (dist_sq >= intens_sq * Constants.LIGHT_DECAY) return Vector3.Zero;
            
            // Create a shadow ray
            var dist = (float)Math.Sqrt(dist_sq);
            light_vec = light_vec / dist;
            var shadow_ray = new Ray(hit.Position + (light_vec * Constants.EPSILON), light_vec);

            var tmp = RayHit.Default();
            foreach (var o in World.Objects) {
                if (o.Intersect(shadow_ray, ref tmp) && tmp.T < dist) {
                    if (debug) DebugData.ShadowRaysOccluded.Add(new Tuple<Ray, RayHit>(shadow_ray, tmp));
                    return new Vector3();
                } 
            }

            // Diffuse
            var materialColor = hit.Obj.Material.Texture == null
                ? hit.Obj.Material.Diffuse
                : hit.Obj.Material.Texture.GetColor(hit.Obj.GetUV(hit));
            var color = materialColor * Math.Max(0.0f, Vector3.Dot(hit.Normal, light_vec));

            // Inverse square law
            var light_powah = light.Intensity / (Constants.PI4 * dist_sq);
            light_powah *= light.AngleEnergy(light_vec.Normalized());

            // Specular will be used separately
            if (hit.Obj.Material.IsGlossy) {
                var hardness = Math.Max(.0f, Vector3.Dot(-ray.Direction, QuickMaths.Reflect(-light_vec, hit.Normal)));
                specular += light.Colour * light_powah * hit.Obj.Material.Specular *
                    (float)Math.Pow(hardness, hit.Obj.Material.Shinyness);
            }

            if (debug) DebugData.ShadowRays.Add(new Tuple<Ray, Vector3>(shadow_ray, ppos));
            return light.Colour * color * light_powah;
        }
       
        /// <summary>
        /// Get diffuse color and specular for the hit point
        /// </summary>
        /// <param name="ray"></param>
        /// <param name="hit"></param>
        /// <param name="specular"></param>
        /// <param name="debug"></param>
        /// <returns></returns>
        protected Vector3 illuminate(Ray ray, RayHit hit, ref Vector3 specular, bool debug)
        {
            var ret = World.Environent.AmbientLight;
            foreach (var light in World.Lights)
            {
                var lPoints = light.GetPoints(Settings.MaxLightSamples, Settings.LSM);
                var localEnergy = Vector3.Zero;
                bool first = true;
                foreach(var lp in lPoints)
                {
                    localEnergy += illuminateBy(lp, light, ray, hit, ref specular, debug && first);
                    first = false;
                }
                localEnergy /= lPoints.Length;
                ret += localEnergy;
            }
            return ret;
        }
        
        protected RayHit intersect(Ray ray)
        {
            var hit = RayHit.Default();
            foreach (var o in World.Objects) o.Intersect(ray, ref hit);
            return hit;
        }
    }
}

File: Fastest Raytracer/raytrace/raytracer.cpp
//
// Created by egordm on 19-5-2018.
//

#include "raytracer.h"
#include "quick_maths.h"

using namespace raytrace;
using namespace raytrace::ray_transform;


void Raytracer::raytrace(World *world, Raster *raster) {
    this->world = world;
    this->raster = raster;
    ddat.primary_rays.clear();
    ddat.shadow_rays.clear();
    ddat.shadow_rays_occluded.clear();
    ddat.refract_rays.clear();

#if MULTI_THREAD
    auto thread_count = std::thread::hardware_concurrency();
    std::vector<std::thread> threads;
    threads.reserve(thread_count);
    auto rpt = RAYTRACE_AREA_HEIGHT / thread_count;
    for (int i = 0; i < thread_count; ++i)
        threads.emplace_back(&Raytracer::render_rows, this, i * rpt, (i + 1) * rpt);
    for (int i = 0; i < thread_count; ++i) threads[i].join();
#else
    render_rows(0, RAYTRACE_AREA_HEIGHT);
#endif

    debug::debug_draw(raster, world, &ddat);
}

void Raytracer::render_rows(uint start, uint end) {
    bool is_debug_row = false;
    uint debug_column = RAYTRACE_DEBUG_FREQUENCY;

    for (uint y = start; y < end; ++y) {
        if (y == RAYTRACE_DEBUG_ROW) is_debug_row = true;

        for (uint x = 0; x < RAYTRACE_AREA_WIDTH; ++x) {
            float wt = (float) x / RAYTRACE_AREA_WIDTH;
            float ht = (float) y / RAYTRACE_AREA_HEIGHT;

            glm::vec3 onPlane = glm::normalize(wt * world->camera.get_fov_plane().nhor +
                                               ht * world->camera.get_fov_plane().nvert +
                                               world->camera.get_fov_plane().origin);

            bool debug = is_debug_row && x == debug_column;
            if (debug) debug_column += RAYTRACE_DEBUG_FREQUENCY;

            auto color = trace_ray(Ray(world->camera.get_position(), onPlane), debug);

            raster->set_pixel(x, y, color);
        }
    }
}

glm::vec3 Raytracer::trace_ray(const Ray &ray, bool debug = false) {
    glm::vec3 ret;
    if (ray.depth > CONSTANT_MAX_DEPTH) return ret;

    RayHit hit = intersect(ray);
    if (hit.object == nullptr) return ret;
    if (debug) ddat.primary_rays.push_back(std::make_tuple(ray, hit));

    glm::vec3 specular;
    glm::vec3 color = illuminate(ray, hit, specular, debug);

    if (hit.object->mat.is_mirror) {
        ret = trace_ray(reflect(ray, hit), debug);
    } else if (hit.object->mat.is_dielectic) {
        auto n1 = ray.is_outside() ? CONSTANT_LIGHT_IOR : hit.object->mat.refraction_index;
        auto n2 = ray.is_outside() ? hit.object->mat.refraction_index : CONSTANT_LIGHT_IOR;
        float reflect_multiplier = maths::fresnel(n1, n2, hit.normal, ray.direction);
        reflect_multiplier = hit.object->mat.reflectivity + (1.f - hit.object->mat.reflectivity) * reflect_multiplier;
        float transmission_multiplier = 1 - reflect_multiplier;

        ret = reflect_multiplier * trace_ray(reflect(ray, hit), debug);
        if (transmission_multiplier > EPSILON) {
            if(hit.object->mat.is_refractive) {
                auto tmp_ray = refract(ray, hit);
                if(!std::isnan(tmp_ray.direction.x))
                    ret += transmission_multiplier * trace_ray_inside(tmp_ray, hit.object, debug);
            } else {
                ret += transmission_multiplier * color;
            }
        }
    } else {
        ret = color;
    }

    return ret + specular;
}

glm::vec3 Raytracer::trace_ray_inside(Ray ray, const Primitive *obj, bool debug) {
    if (ray.is_outside()) return trace_ray(ray, debug); // Should not happen anyway
    glm::vec3 ret;
    float multiplier = 1;
    float absorb_distance = 0;

    RayHit hit;
    while (ray.depth < CONSTANT_MAX_DEPTH - 1 && multiplier > EPSILON) {
        if (!obj->intersect(ray, hit))
            return ret; // Should not happen either
        if(debug) ddat.refract_rays.push_back(std::make_tuple(ray, hit));

        // Beer absorption
        absorb_distance += hit.t;
        auto absorb = glm::exp(-obj->mat.absorb * absorb_distance);

        auto reflect_multiplier = maths::fresnel(obj->mat.refraction_index, CONSTANT_LIGHT_IOR, ray.direction, hit.normal);
        auto refract_multiplier = 1.f - reflect_multiplier;

        if (refract_multiplier > EPSILON)
            ret += trace_ray(refract(ray, hit), debug) * refract_multiplier * multiplier * absorb;

        ray = reflect(ray, hit);

        multiplier *= reflect_multiplier;
    }

    return ret;
}

glm::vec3 Raytracer::illuminate_by(const Light *light, const Ray &ray, const RayHit &hit, glm::vec3 &specular,
                                   bool debug = false) {
    auto light_vec = light->get_position() - hit.position;
    float dist_sq = glm::dot(light_vec, light_vec);
    float intens_sq = powf(light->get_intensity(), 2);
    if (dist_sq >= intens_sq * LIGHT_DECAY) return glm::vec3();

    auto dist = sqrtf(dist_sq);
    light_vec = light_vec / dist;
    raytrace::Ray occlusion_ray(hit.position + (light_vec * EPSILON), light_vec);

    RayHit tmp;
    tmp.t = dist;
    for (auto object : world->objects) {
        if (object->intersect(occlusion_ray, tmp)) {
            if (debug) ddat.shadow_rays_occluded.push_back(std::make_tuple(occlusion_ray, tmp));
            return glm::vec3();
        }
    }

    // Diffuse
    glm::vec3 color = hit.object->mat.diffuse * fmaxf(0.0f, glm::dot(hit.normal, light_vec));

    // Specular
    if (hit.object->mat.glossy) {
        auto hardness = fmaxf(.0f, glm::dot(-ray.direction, glm::reflect(-light_vec, hit.normal)));
        specular += light->get_color() * light->get_intensity() * hit.object->mat.specular *
                    powf(hardness, hit.object->mat.shinyness);
    }

    if (debug) ddat.shadow_rays.push_back(std::make_tuple(occlusion_ray, light->get_position()));
    return light->get_color() * color;
}

glm::vec3 Raytracer::illuminate(const Ray &ray, const RayHit &hit, glm::vec3 &specular, bool debug = false) {
    glm::vec3 ret;

    for (const auto &light : world->lights) ret += illuminate_by(light, ray, hit, specular, debug);

    return ret;
}

RayHit Raytracer::intersect(const Ray &ray) {
    RayHit hit = raytrace::RayHit();
    for (auto object : world->objects) object->intersect(ray, hit);
    return hit;
}
